File: MATLAB/Solvers/SPSA/SPSA.m
%==========================================================================
%                       The SPSA Algorithm
%==========================================================================
% DATE
%        Feb 2017 (Updated in Dec 2019 by David Eckman)
%
% AUTHOR
%        Anna Dong
%
% REFERENCE	
%        James C. Spall, (1998). An Overview of the Simultaneous 
%        Perturbation Method for Efficient Optimization.
%
%==========================================================================
%
% INPUT
%        problem
%              Problem function name
%        probstructHandle
%              Problem structure function name 
%        problemRng
%              Random number generators (streams) for problems
%        solverRng
%              Random number generator (stream) for solver
%
% OUTPUT
%        Ancalls
%              An array (size = 'NumSoln' X 1) of budget expended
%        A
%              An array (size = 'NumSoln' X 'dim') of solutions
%              returned by solver
%        AFnMean
%              An array (size = 'NumSoln' X 1) of estimates of expected
%              objective function value
%        AFnVar
%              An array of variances corresponding to
%              the objective function at A
%              Equals NaN if solution is infeasible
%        AFnGrad
%              An array of gradient estimates at A; not reported
%        AFnGradCov
%              An array of gradient covariance matrices at A; not reported
%        AConstraint
%              A vector of constraint function estimators; not applicable
%        AConstraintCov
%              An array of covariance matrices corresponding to the
%              constraint function at A; not applicable
%        AConstraintGrad
%              An array of constraint gradient estimators at A; not
%              applicable
%        AConstraintGradCov
%              An array of covariance matrices of constraint gradient
%              estimators at A; not applicable
%
%==========================================================================

%% Simultaneous Perturbation Stochastic Approximation (SPSA)
function [Ancalls, A, AFnMean, AFnVar, AFnGrad, AFnGradCov, ...
    AConstraint, AConstraintCov, AConstraintGrad, ...
    AConstraintGradCov] = SPSA(probHandle, probstructHandle, problemRng, ...
    solverRng)

%% Unreported
AFnGrad = NaN;
AFnGradCov = NaN;
AConstraint = NaN;
AConstraintCov = NaN;
AConstraintGrad = NaN;
AConstraintGradCov = NaN;

% Separate the two solver random number streams
solverInitialRng = solverRng{1}; % RNG for finding initial solutions
solverInternalRng = solverRng{2}; % RNG for the solver's internal randomness

% Set default values
sensitivity = 10^(-7); % shrinking scale for VarBds
alpha = 0.602;
gamma = 0.101;
step = 0.1; % 'What is the initial desired magnitude of change in the theta elements?'
gavg = 1; % 'How many averaged SP gradients will be used per iteration? '
r = 30; % Number of replications takes at each solution
NL = 2; % 'How many loss function evaluations do you want to use in this gain calculation? '
% NL/(2*gavg) SP gradient estimates

% Get initial information about the problem
[minmax, dim, ~, ~, VarBds, ~, ~, ~, budget, ~, ~, ~] = probstructHandle(0);

% Shrink VarBds to prevent floating point errors
VarBds(:,1) = VarBds(:,1) + sensitivity;
VarBds(:,2) = VarBds(:,2) - sensitivity;

% Determine maximum number of solutions that can be sampled within max budget
MaxNumSoln = floor(budget/r); 

% Initialize larger than necessary (extra point for end of budget)
Ancalls = zeros(MaxNumSoln + 1, 1);
A = zeros(MaxNumSoln + 1, dim);
AFnMean = zeros(MaxNumSoln + 1, 1);
AFnVar = zeros(MaxNumSoln + 1, 1);

% Using CRN: for each solution, start at substream 1
problemseed = 1;
% If not using CRN: problemseed needs to be updated throughout the code

% Get initial solution
RandStream.setGlobalStream(solverInitialRng);
[~, ~, ~, ~, ~, ~, ~, theta0, ~, ~, ~, ~] = probstructHandle(1);

% Evaluate theta0
[ftheta, fthetaVar, ~, ~, ~, ~, ~, ~] = probHandle(theta0, r, problemRng, problemseed);
Bspent = r;
% fthetaVar answers 'What is the standard deviation of the measurement noise at i.c.? '
c = max(fthetaVar/gavg^0.5, .0001);

% Record initial solution data
Ancalls(1) = 0; 
A(1,:) = theta0;
AFnMean(1) = ftheta;
AFnVar(1) = fthetaVar;

% Record only when recommended solution changes
record_index = 2;

% Relabel initial solution and mark its estimate as best so far
theta = theta0;
ftheta_best = ftheta;

% Set other parameters
nEvals = round((budget/r)/3 *2); % 'What is the expected number of loss evaluations per run? '
Aalg = .10*nEvals/(2*gavg);

% Determine initial value for the parameter a (according to Section III.B of Spall (1998)) 
gbar = zeros(1, dim);

for i = 1:NL/(2*gavg)
   ghat = zeros(1, dim);
   for j = 1:gavg
      % Generate a random random direction (delta)
      RandStream.setGlobalStream(solverInternalRng);
      delta = 2*round(rand(1, dim)) - 1;

      % Determine points forward/backward relative to random direction
      thetaplus = theta + c*delta;
      thetaminus = theta - c*delta;
      thetaplus2 = checkCons(VarBds, thetaplus, theta);
      thetaminus2 = checkCons(VarBds, thetaminus, theta);

      % Evaluate two points and update budget spent
      [fn, ~] = probHandle(thetaplus2, r, problemRng, problemseed); 
      yplus = -minmax*fn;
      [fn, ~] = probHandle(thetaminus2, r, problemRng, problemseed); 
      yminus = -minmax*fn; 
      Bspent = Bspent + 2*r;

      % Estimate gradient
      ghat = (yplus - yminus)./(2*c*delta) + ghat;
   end
   gbar = gbar + abs(ghat/gavg);
end

meangbar = mean(gbar);
meangbar = meangbar/(NL/(2*gavg));
a = step*((Aalg + 1)^alpha)/meangbar;

% Run the main algorithm

k = 1; % Iteration counter

while Bspent <= budget
    
    ak = a/(k + Aalg)^alpha;
    ck = c/(k^gamma);
    
    % Generate a random random direction (delta)
    RandStream.setGlobalStream(solverInternalRng);
    delta = 2*round(rand(1, dim)) - 1;

    % Determine points forward/backward relative to random direction
    thetaplus = theta + ck*delta;
    thetaminus = theta - ck*delta;
    thetaplus2 = checkCons(VarBds, thetaplus, theta);
    thetaminus2 = checkCons(VarBds, thetaminus, theta);

    % Evaluate two points and update budget spent
    [fn, ~] = probHandle(thetaplus2, r, problemRng, problemseed); 
    yplus = -minmax*fn;
    [fn, ~] = probHandle(thetaminus2, r, problemRng, problemseed); 
    yminus = -minmax*fn;
    Bspent = Bspent + 2*r;
    
    % Estimate current solution's obj fun value by averaging 
    % estimates at thetaplus2 and thetaminus2
    ftheta = -minmax*(yplus + yminus)/2;
    
    % Check if current solution looks better than the best so far
    if -minmax*ftheta < -minmax*ftheta_best && Bspent <= budget
        
        theta_best = theta;
        ftheta_best = ftheta;
        
        %Record data from the new best solution
        Ancalls(record_index) = Bspent;
        A(record_index,:) = theta_best;
        AFnMean(record_index) = ftheta_best;
        AFnVar(record_index) = NaN; % no estimate of variance
        record_index = record_index + 1;
        
    end

    % Estimate gradient
    ghat = (yplus - yminus)./(2*c*delta);

    % Take step and check feasibility
    theta_next = theta - ak*ghat;
    theta = checkCons(VarBds, theta_next, theta);
    
end

% Record solution at max budget
Ancalls(record_index) = budget;
A(record_index,:) = A(record_index - 1,:);
AFnMean(record_index) = AFnMean(record_index - 1);
AFnVar(record_index) = AFnVar(record_index - 1);

% Trim empty rows from data
Ancalls = Ancalls(1:record_index);
A = A(1:record_index,:);
AFnMean = AFnMean(1:record_index);
AFnVar = AFnVar(1:record_index);

%% Helper Functions
% Helper 1: Check & Modify (if needed) the new point, based on VarBds.
% ssolsV2 original solution. ssolsV expected solution.
    function modiSsolsV = checkCons(VarBds, ssolsV, ssolsV2)
        col = size(ssolsV, 2);
        stepV = ssolsV - ssolsV2;
        % t>0 for the correct direction
        tmaxV = ones(2, col);
        uV = VarBds(stepV > 0, 2); uV = uV';
        lV = VarBds(stepV < 0, 1); lV = lV';
        if isempty(uV) == 0 %length(uV)> 0
            tmaxV(1, stepV > 0) = (uV - ssolsV2(stepV > 0)) ./ stepV(stepV > 0);
        end
        if isempty(lV) == 0 %length(lV)>0
            tmaxV(2,stepV < 0) = (lV - ssolsV2(stepV < 0)) ./ stepV(stepV < 0);
        end
        t = min(min(tmaxV));
        modiSsolsV = ssolsV2 + t*stepV;
        %rounding error, may remove
        for kc = 1:col
            if abs(modiSsolsV(kc)) < 0.00000005
                modiSsolsV(kc) = 0;
            end
        end
    end

end

File: simopt/solvers/SPSA.py
""" 
Summary

Simultaneous perturbation stochastic approximation (SPSA) is an algorithmic method for optimizing systems with multiple unknown parameters.

"""

from random import random
from base import Solver
from base import random
import numpy as np

class SPSA(solver):
    """
    Simultaneous perturbation stochastic approximation (SPSA) is an algorithmic method for optimizing systems with multiple unknown parameters.
    ----------
    name : string
        name of solver
    objective_type : string
        description of objective types:
            "single" or "multi"
    constraint_type : string
        description of constraints types:
            "unconstrained", "box", "deterministic", "stochastic"
    variable_type : string
        description of variable types:
            "discrete", "continuous", "mixed"
    gradient_needed : bool
        indicates if gradient of objective function is needed
    factors : dict
        changeable factors (i.e., parameters) of the solver
    specifications : dict
        details of each factor (for GUI, data validation, and defaults)
    rng_list : list of rng.MRG32k3a objects
        list of RNGs used for the solver's internal purposes
    Arguments
    ---------
    name : str
        user-specified name for solver
    fixed_factors : dict
        fixed_factors of the solver
    See also
    --------
    base.Solver
    """
    def __init__(self, name="SPSA", fixed_factors={}):
            self.name = name
            self.objective_type = "single"                  ### Need to update these
            self.constraint_type = "stochastic"
            self.variable_type = "mixed"
            self.gradient_needed = False
            self.specifications = {
                "sensititvity": {
                    "description": "shrinking scale for VarBds",
                    "datatype": float,
                    "default": 10^(-7)
                },
                "alpha": {
                    "description": "Non negative coefficient",
                    "datatype": float,
                    "default": 0.602
                },
                "gamma": {
                    "description": "Non negative coefficient",
                    "datatype": float,
                    "default": 0.101
                },
                "step": {
                    "description": "The initial desired magnitude of change in the theta elements",
                    "datatype": float,
                    "default": 0.1
                },
                "gavg": {
                    "description": "How many averaged SP gradients will be used per iteration?",
                    "datatype": float,
                    "default": 1
                },
                "r": {
                    "description": "Number of replications takes at each solution",   #increased r means more accurate gradient
                    "datatype": int,
                    "default": 30
                },
                "NL": {
                    "description": "How many loss function evaluations do you want to use in this gain calculation?",
                    "datatype": int,
                    "default": 2
                }
            }
            self.check_factor_list = {
                "sensititvity": self.check_sensititvity,
                "alpha": self.check_alpha,
                "gamma": self.check_gamma,
                "step": self.check_step,
                "gavg": self.check_gavg,
                "r": self.check_r,
                "NL": self.check_NL
            }
            super().__init__(fixed_factors)

    def check_sensititvity(self):
        return self.factors["sensititvity"] > 0

    def check_alpha(self):
        return self.factors["alpha"] > 0

    def gamma(self):
        return self.factors["check_gamma"] > 0

    def step(self):
        return self.factors["check_step"] > 0

    def check_gavg(self):
        return self.factors["gavg"] > 0

    def check_r(self):
        return self.factors["r"] > 0

    def check_NL(self):
        return self.factors["NL"] > 0

    def gen_simul_pert_vec(self, inital_solution):
        """Generates a new simulatanious pertubation vector to be applied, uses input of inital solution to get vector length"""
        
        SP_vect = []
        
        for i in inital_solution:
            delta = random.choices([-1,1], [.5,.5])
            SP_vect.append(delta)
        
        return SP_vect

    def solve(self, problem):
        """
        Run a single macroreplication of a solver on a problem.
        Arguments
        ---------
        problem : Problem object
            simulation-optimization problem to solve
        
>>      Should I have anything else in this section?

        Returns
        -------
        Reccomended_solutions

        Intermediate_budgets 
        
        #loss function is embedded in the solution object
        
        Anncalls : Array
            An array (size = 'NumSoln' X 1) of budget expended
        A : Array
             An array (size = 'NumSoln' X 'dim') of solutions returned by solver
        AFnMean : Array
            An array (size = 'NumSoln' X 1) of estimates of expected objective function value
        AFnVar : Array 
            An array of variances corresponding to the objective function at A 
            Equals NaN if solution is infeasible
        AFnGrad : Array
            An array of gradient estimates at A; not reported
        AFnGradCov : Array 
            An array of gradient covariance matrices at A; not reported
        AConstraint : Vector
            A vector of constraint function estimators; not applicable
        AConstraintCov : Arrary
            An array of covariance matrices corresponding to the constraint function at A; not applicable
        AConstraintGrad : Array
            An array of constraint gradient estimators at A; not applicable
        AConstraintGradCov: Array
            An array of covariance matrices of constraint gradient estimators at A; not applicable
        """
        recommended_solns = []
        intermediate_budgets = []
        expended_budget = 0
        ghat = [0,0]
        c = max(fthetaVar/self.factors["gavg"]^0.5,.0001)   #check line 113 in matlab code, need to determine fthetavar
        A = 1 + int(problem.factors["budget"]/self.factors["r"])
        k = 0
        # Designate random number generator for random sampling.
        find_next_soln_rng = self.rng_list[2] #the rng list is automatically integrated into the .random class
        # Sequentially generate random solutions and simulate them.
        while expended_budget < problem.factors["budget"]:
            if expended_budget == 0:
                # Start at initial solution and record as best.
                new_x = problem.factors["initial_solution"] #CRNs = Common Random Numbers, not what is used here, need a specic RN Stream #3, Look at line 97 on RandSearch
                new_solution = new_x #function of a solver super class under base.py
                best_solution = new_solution                            #Can use random.choices([-1,1], .5,.5)
                recommended_solns.append(new_solution)
                intermediate_budgets.append(expended_budget)
            else:
                #generate Simulanious Pertubation Vector
                k += 1
                delta_k = self.gen_simul_pert_vec(new_solution) #outputs either +tv or -tv integers aka +1 or -1 for every variable 
                #Loss function evaluation
                ck = c/(k^self.factors(["gamma"]))
                ak = A/(k + )
                #loss func 1
                L1 = np.dot(new_solution,np.add(delta_k,np.multiply(ck,delta_k)))
                #loss func 2
                L2 = np.dot(new_solution,np.subtract(delta_k,np.multiply(ck,delta_k)))
                #Gradient Approximation
                ghat = np.dot(np.divide(L1-L2, 2 * ck), np.transform(delta_k))
                #Updating previous solution estimate
                new_solution = new_solution - ghat
                recommended_solns.append(new_solution)
                #iteration or termnination
                
                
                
            # Also check for feasibility w.r.t. stochastic constraints.
            if (problem.minmax * new_solution.objectives_mean
                    > problem.minmax * best_solution.objectives_mean and
                    all(new_solution.stoch_constraints_mean[idx] >= 0 for idx in range(problem.n_stochastic_constraints))):
                # If better, record incumbent solution as best.
                best_solution = new_solution
                recommended_solns.append(new_solution)
                intermediate_budgets.append(expended_budget)
        return recommended_solns, intermediate_budgets
